### ISSUE TITLE
Add support for Subscriptions

### ISSUE THREAD

**Issue by Joezo** [nearform/graphql-hooks#72]
### Description

We support Queries and Mutations but don't support the third operation type, Subscriptions. We should look at adding that.

### Suggested implementation
I would guess websockets?

**Comment by mcollina** [comment]
Subscriptions will be extraordinarly complicated. I would restrain from using (and implementing) them, and they are not really part of the spec (See https://github.com/facebook/graphql/blob/master/rfcs/Subscriptions.md, it's a draft).

On the server side, they would scale _horribly_. I've worked on pubsub systems, and this type of unbounded updates is extremely hard to use in production.

It would be fine to experiment with those, but as a separate module.

**Comment by Joezo** [comment]
I think you're right @mcollina - it does look incredibly complicated. Might be worth keeping this around as a long running issue to track subscription discussion.

**Comment by brainkim** [comment]
> On the server side, they would scale horribly. I've worked on pubsub systems, and this type of unbounded updates is extremely hard to use in production.

@mcollina Genuinely curious to hear more on this. What do you mean by “unbounded” updates? Are `apollographql/graphql-subscriptions` and packages which implement the `PubSubEngine` interface fatally flawed?

**Comment by mcollina** [comment]
It closely depends on how you code the resolvers: it's extremely easy to shoot yourself in the foot.

The following example is essentially flowed:

```js
const SOMETHING_CHANGED_TOPIC = 'something_changed';

export const resolvers = {
  Subscription: {
    somethingChanged: {
      subscribe: () => withFilter(
        (_, args) => pubsub.asyncIterator(SOMETHING_CHANGED_TOPIC),
        (payload, variables) => payload.anAttribute === variables.AnAttribute,
      ),
  },
}
```

Essentially you are doing the vast majority of the filtering on your Node.js processes. This __will not scale__, because the processes will receive the vast majority of the updates. You need to use different topics and _never ever filter_ within a resolver. At some point, the filtering will use 100% of the CPU of every process without sending any data. Doing filtering within the resolvers (as all example do) works well at small scale but is fatally flawed on a huge volume of messages moved. 

The whole purpose of GraphQL is doing advanced filtering. Not be able to do filtering makes it a non compelling technology compared to alternatives for real-time communication.

**Comment by good-idea** [comment]
What would make this difficult on the client?

**Comment by brainkim** [comment]
@mcollina Had an aha moment today after reading your comment. To rephrase what you’re saying for my own understanding:

Pubsub services are used to scale horizontally. However, if every graphql resolver subscribes to broad hard-coded topics like "something changed" and filters payloads in memory, each process ends up receiving virtually every message that the service publishes, effectively negating any scaling advantages as the number of messages grows. This is a fundamental problem with graphql subscriptions because while graphql’s main advantage is its ability to arbitrarily filter responses based on client needs, most pubsub services rely on filtering based on topics i.e. simple string keys, which leads to a mismatch between graphql and pubsub services.

Thanks for your explanation!

**Comment by mcollina** [comment]
I'm going to partially change my mind. While subscriptions implemented as "backend" services, backed by a pubsub, are hard to implement correctly, subscriptions can be implemented effeciently at the database layer, e.g. with https://hasura.io/. Maybe we should give it a try.

**Comment by bmullan91** [comment]
Fixed in `graphql-hooks@4.2.0` 🚀
